src/backer/cardaning.py

```diff
@@ -26,9 +24,9 @@
 FUNDING_AMOUNT = os.environ.get('FUNDING_AMOUNT', 30000000)
 TRANSACTION_AMOUNT = os.environ.get('TRANSACTION_AMOUNT', 1000000)
 MIN_BLOCK_CONFIRMATIONS = os.environ.get('MIN_BLOCK_CONFIRMATIONS', 3)
-SHELLY_UNIX = os.environ.get('SHELLY_UNIX', 1666656000) #"2022-10-25T00:00:00Z"
 TRANSACTION_SECURITY_DEPTH = os.environ.get('TRANSACTION_SECURITY_DEPTH', 16)
 TRANSACTION_TIMEOUT_DEPTH = os.environ.get('TRANSACTION_TIMEOUT_DEPTH', 32)
+MAX_TRANSACTION_SIZE = 16380 # 16kB - 4 bytes, Somtimes Ogmios create to a bigger transaction so we decrease 4 bytes to pass that
```

Comment:
We are using pycardano to build and sign the transaction, And the size of tx is exactly the len of tx cbor from pycardano,
The pycardano tx after sign also contains UTXOs, signatures, fee, ...
And the pycardano also validate the transaction size itself, so I am making some test to consider that we can use the exception which returned from pycardano to making our "event batch" tx like this:
 
```
                    try:
                        signed_tx = builder.build_and_sign([self.payment_signing_key],
                                                        change_address=self.spending_addr,
                                                        merge_change=True)
                        temp_tx_cbor = signed_tx.to_cbor()
                    except pycardano.exception.InvalidTransactionException as e:
                        if "exceeds the max limit" in str(e):    
                            # Stop read event array and submit the event batch                        
                            break
```

Reply:
The size of the tx unsigned contains the fee and UTXOs, but yes the signatures added on after are also included in the final size. But we shouldn't remove 4 bytes here.

We should just use the full 16kB and check after generating the fully signed tx, I think it's easier. We can optimise in the future.

But great, if pycardano validates the size, then we can just use that. So if there's more KELs to add, we add them until we hit this exception and go back one step. Not perfect, but again we can optimise in the future rather than making risky calculations.

Reply:
>But we shouldn't remove 4 bytes here
Yes, I am making some test to make sure, It's better we must use exactly 16kB
> we add them until we hit this exception and go back one step
I prefer this now
> Not perfect, but again we can optimise in the future rather than making risky calculations
Now We cannot create a transaction larger than 16kB because the pycardano does not allow that

Reply:
I test a transaction with exactly 16384 bytes, that tx pass to pycardano (pycardano still recognize 16384 bytes) then the pycardano pass to python ogmios client, then ogmios client send to ogmios server (this is implemented by Haskell) => now ogmios server response that the tx size is 16387 bytes. This is the response:
Ogmios responded with error: {'jsonrpc': '2.0', 'method': 'submitTransaction', 'error': {'code': 3119, 'message': "The transaction exceeds the maximum size allowed by the protocol. Indeed, once serialized, transactions must be under a bytes limit specified by a protocol parameter. The field 'data.measuredTransactionSize' indicates the actual measured size of your serialized transaction, whereas 'data.maximumTransactionSize' indicates the current maximum size enforced by the ledger.", 'data': {'measuredTransactionSize': {'bytes': 16387}, 'maximumTransactionSize': {'bytes': 16384}}}, 'id': None}
I guess that the problem come from CBOR Encoding/Decoding between python and haskell so I prefer to add a "Adjust Safety Margin" for that, It maybe 3 bytes or 4 bytes or 16 bytes, ...

Reply:
agreed on slack